File: app/schemas/nutrition.py
"""Pydantic models for nutrition analysis.

These are the validation / LLM-I/O layer. ``MacroEstimate`` is the strict
structured-output schema the vision model must return.
"""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class FoodItem(BaseModel):
    name: str
    quantity: str | None = Field(
        default=None, description="e.g. '1 cup', 'approx 150g'"
    )
    plate_share: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
        description=(
            "Fraction of the food's own vessel this item fills (0-1). Items "
            "sharing one plate should sum to ~1.0; a soup in its own bowl is 1.0."
        ),
    )
    calories: float = 0
    protein_g: float = 0
    carbs_g: float = 0
    fat_g: float = 0
    fatsecret_food_id: str | None = None
    fatsecret_match: str | None = Field(
        default=None, description="FatSecret food_description for this match"
    )


class MacroEstimate(BaseModel):
    """Structured result of a food photo / description analysis."""

    items: list[FoodItem] = Field(default_factory=list)
    calories: float = 0
    protein_g: float = 0
    carbs_g: float = 0
    fat_g: float = 0
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    food_confidence: float = Field(
        default=0.0, ge=0.0, le=1.0, description="Certainty about WHAT the food is"
    )
    portion_confidence: float = Field(
        default=0.0, ge=0.0, le=1.0, description="Certainty about HOW MUCH there is"
    )
    assumptions: list[str] = Field(default_factory=list)
    description: str = ""

    def is_low_confidence(self, threshold: float = 0.5) -> bool:
        return self.confidence < threshold

    def needs_portion_confirm(self, threshold: float = 0.6) -> bool:
        return self.portion_confidence < threshold

    def apply_multiplier(self, factor: float) -> "MacroEstimate":
        """Return a copy with all macros (and per-item macros) scaled."""
        scaled_items = [
            item.model_copy(
                update={
                    "calories": round(item.calories * factor, 1),
                    "protein_g": round(item.protein_g * factor, 1),
                    "carbs_g": round(item.carbs_g * factor, 1),
                    "fat_g": round(item.fat_g * factor, 1),
                }
            )
            for item in self.items
        ]
        return self.model_copy(
            update={
                "items": scaled_items,
                "calories": round(self.calories * factor, 1),
                "protein_g": round(self.protein_g * factor, 1),
                "carbs_g": round(self.carbs_g * factor, 1),
                "fat_g": round(self.fat_g * factor, 1),
            }
        )

    def totals_line(self) -> str:
        return (
            f"{round(self.calories)} kcal · "
            f"P {round(self.protein_g)}g · "
            f"C {round(self.carbs_g)}g · "
            f"F {round(self.fat_g)}g"
        )


class ImageRef(BaseModel):
    """Lightweight, persisted reference to an image kept on Telegram."""

    file_id: str
    file_unique_id: str
    width: int | None = None
    height: int | None = None
    taken_at: datetime
    caption: str | None = None


class DailyTotals(BaseModel):
    date: str
    calories: float = 0
    protein_g: float = 0
    carbs_g: float = 0
    fat_g: float = 0
    meals: int = 0
    workouts: int = 0
    workout_minutes: int = 0

    target_calories: int | None = None
    target_protein_g: int | None = None
    target_carbs_g: int | None = None
    target_fat_g: int | None = None

File: src/schemas/nutrition.ts
import { z } from "zod";

export const containerTypeSchema = z.enum([
  "cup",
  "plate",
  "bowl",
  "packaged",
  "unknown",
]);

export const portionSchema = z.object({
  container_type: containerTypeSchema.default("unknown"),
  container_volume_ml: z.number().nullable().optional(),
  fill_fraction: z.number().default(1),
  notes: z.string().nullable().optional(),
});

export const foodItemSchema = z.object({
  name: z.string(),
  quantity: z.string().nullable().optional(),
  plate_share: z.number().nullable().optional(),
  weight_g: z.number().default(0),
  volume_ml: z.number().nullable().optional(),
  volume_fraction: z.number().nullable().optional(),
  calories: z.number().default(0),
  protein_g: z.number().default(0),
  carbs_g: z.number().default(0),
  fat_g: z.number().default(0),
});

export const macroEstimateSchema = z.object({
  items: z.array(foodItemSchema).default([]),
  portion: portionSchema.optional(),
  calories: z.number().default(0),
  protein_g: z.number().default(0),
  carbs_g: z.number().default(0),
  fat_g: z.number().default(0),
  confidence: z.number().default(0),
  food_confidence: z.number().default(0),
  portion_confidence: z.number().default(0),
  assumptions: z.array(z.string()).default([]),
  description: z.string().default(""),
});

export type PortionEstimate = z.infer<typeof portionSchema>;
export type FoodItem = z.infer<typeof foodItemSchema>;
export type MacroEstimate = z.infer<typeof macroEstimateSchema>;

/** Strict schema for LLM structured output (all macro fields required). */
export const strictFoodItemSchema = z.object({
  name: z.string(),
  quantity: z.string().nullable().optional(),
  plate_share: z.number().nullable().optional(),
  weight_g: z.number(),
  volume_ml: z.number().nullable().optional(),
  volume_fraction: z.number().nullable().optional(),
  calories: z.number(),
  protein_g: z.number(),
  carbs_g: z.number(),
  fat_g: z.number(),
});

export const strictPortionSchema = z.object({
  container_type: containerTypeSchema,
  container_volume_ml: z.number().nullable(),
  fill_fraction: z.number(),
  notes: z.string().nullable(),
});

export const clarifyToggleSchema = z.object({
  id: z.string(),
  label: z.string(),
});

export const clarifyExclusiveOptionSchema = z.object({
  id: z.string(),
  label: z.string(),
});

export const clarifyExclusiveSchema = z.object({
  id: z.string(),
  prompt: z.string(),
  options: z.array(clarifyExclusiveOptionSchema),
});

export const clarificationSpecSchema = z.object({
  toggles: z.array(clarifyToggleSchema).default([]),
  exclusive: clarifyExclusiveSchema.nullable().default(null),
});

export type ClarifyToggle = z.infer<typeof clarifyToggleSchema>;
export type ClarifyExclusive = z.infer<typeof clarifyExclusiveSchema>;
export type ClarificationSpec = z.infer<typeof clarificationSpecSchema>;

export const strictMacroEstimateSchema = z.object({
  items: z.array(strictFoodItemSchema),
  portion: strictPortionSchema,
  calories: z.number(),
  protein_g: z.number(),
  carbs_g: z.number(),
  fat_g: z.number(),
  confidence: z.number(),
  food_confidence: z.number(),
  portion_confidence: z.number(),
  assumptions: z.array(z.string()),
  description: z.string(),
});

/** Strict schema for vision LLM output (includes clarification spec). */
export const strictVisionOutputSchema = strictMacroEstimateSchema.extend({
  clarification: clarificationSpecSchema,
});

export type VisionOutput = z.infer<typeof strictVisionOutputSchema>;

/** Strip vision-only clarification before persisting a meal estimate. */
export function stripClarificationFromEstimate(
  raw: MacroEstimate & { clarification?: ClarificationSpec | null },
): MacroEstimate {
  const { clarification: _clarification, ...estimate } = raw;
  return estimate;
}

function roundMacro(n: number): number {
  return Math.round(n * 10) / 10;
}

function sumItemMacros(items: FoodItem[]): Pick<
  MacroEstimate,
  "calories" | "protein_g" | "carbs_g" | "fat_g"
> {
  return items.reduce(
    (acc, item) => ({
      calories: roundMacro(acc.calories + item.calories),
      protein_g: roundMacro(acc.protein_g + item.protein_g),
      carbs_g: roundMacro(acc.carbs_g + item.carbs_g),
      fat_g: roundMacro(acc.fat_g + item.fat_g),
    }),
    { calories: 0, protein_g: 0, carbs_g: 0, fat_g: 0 },
  );
}

function scaleItemFields(item: FoodItem, factor: number): FoodItem {
  if (factor >= 0.999 && factor <= 1.001) return item;
  const scale = (n: number) => roundMacro(n * factor);
  return {
    ...item,
    weight_g: scale(item.weight_g),
    volume_ml:
      item.volume_ml != null ? scale(item.volume_ml) : item.volume_ml,
    calories: scale(item.calories),
    protein_g: scale(item.protein_g),
    carbs_g: scale(item.carbs_g),
    fat_g: scale(item.fat_g),
  };
}

/** Cap liquid volumes to container capacity; recompute meal totals from items. */
export function normalizePortionEstimate(raw: MacroEstimate): MacroEstimate {
  const estimate: MacroEstimate = {
    ...raw,
    portion: raw.portion
      ? {
          container_type: raw.portion.container_type ?? "unknown",
          container_volume_ml: raw.portion.container_volume_ml ?? null,
          fill_fraction: raw.portion.fill_fraction ?? 1,
          notes: raw.portion.notes ?? null,
        }
      : undefined,
    items: (raw.items ?? []).map((item) => ({
      ...item,
      weight_g: item.weight_g ?? 0,
    })),
    assumptions: [...(raw.assumptions ?? [])],
  };

  const portion = estimate.portion;
  if (!portion?.container_volume_ml || portion.container_volume_ml <= 0) {
    if (estimate.items.length > 0) {
      const totals = sumItemMacros(estimate.items);
      return { ...estimate, ...totals };
    }
    return estimate;
  }

  const maxLiquidMl = roundMacro(
    portion.container_volume_ml * (portion.fill_fraction ?? 1),
  );
  if (maxLiquidMl <= 0) return estimate;

  let totalLiquidMl = 0;
  for (const item of estimate.items) {
    if (item.volume_ml != null && item.volume_ml > 0) {
      totalLiquidMl += item.volume_ml;
    }
  }

  const volumeFractionSum = estimate.items.reduce(
    (sum, item) => sum + (item.volume_fraction ?? 0),
    0,
  );
  if (volumeFractionSum > 1.05) {
    estimate.portion_confidence = Math.min(estimate.portion_confidence, 0.4);
    if (
      !estimate.assumptions.some((a) =>
        a.startsWith("portion: liquid fractions exceed container"),
      )
    ) {
      estimate.assumptions.push(
        "portion: liquid fractions exceed container",
      );
    }
  }

  if (totalLiquidMl > maxLiquidMl && totalLiquidMl > 0) {
    const factor = maxLiquidMl / totalLiquidMl;
    estimate.items = estimate.items.map((item) => {
      if (item.volume_ml == null || item.volume_ml <= 0) return item;
      return scaleItemFields(item, factor);
    });
    if (
      !estimate.assumptions.some((a) =>
        a.startsWith("portion: capped liquid volume to container"),
      )
    ) {
      estimate.assumptions.push(
        `portion: capped liquid volume to container (${maxLiquidMl} ml)`,
      );
    }
  }

  const totals = sumItemMacros(estimate.items);
  return { ...estimate, ...totals };
}

export function normalizeMacroEstimate(raw: MacroEstimate): MacroEstimate {
  const withPortion = normalizePortionEstimate(raw);
  const estimate = { ...withPortion };
  if (estimate.food_confidence || estimate.portion_confidence) {
    estimate.confidence = Math.min(
      estimate.food_confidence || 1,
      estimate.portion_confidence || 1,
    );
  }
  return estimate;
}

export function macroEstimateFromDict(payload: unknown): MacroEstimate {
  const parsed = macroEstimateSchema.parse(payload);
  return normalizeMacroEstimate(parsed);
}

export function needsPortionConfirm(
  estimate: MacroEstimate,
  threshold: number,
): boolean {
  return estimate.portion_confidence < threshold;
}

export function applyMultiplier(
  estimate: MacroEstimate,
  factor: number,
): MacroEstimate {
  const scale = (n: number) => Math.round(n * factor * 10) / 10;
  return {
    ...estimate,
    portion: estimate.portion
      ? {
          ...estimate.portion,
          container_volume_ml:
            estimate.portion.container_volume_ml != null
              ? scale(estimate.portion.container_volume_ml)
              : null,
        }
      : undefined,
    items: estimate.items.map((item) => ({
      ...item,
      weight_g: scale(item.weight_g),
      volume_ml: item.volume_ml != null ? scale(item.volume_ml) : item.volume_ml,
      calories: scale(item.calories),
      protein_g: scale(item.protein_g),
      carbs_g: scale(item.carbs_g),
      fat_g: scale(item.fat_g),
    })),
    calories: scale(estimate.calories),
    protein_g: scale(estimate.protein_g),
    carbs_g: scale(estimate.carbs_g),
    fat_g: scale(estimate.fat_g),
  };
}
